### ISSUE TITLE
The wrap option in diff mode

### ISSUE THREAD

**Issue by rickhowe** [vim/vim#6639]
I am enjoying to use diff mode.
But I am wondering why the wrap option always forced to be off.

I know its reason is to properly align text between diff windows,
but it is more important for some users (including me) to always wrap lines and see all text.
We have to reset the option every time in diff windows (or set autocmd).

Is it possible and reasonable to remain the wrap option in diff mode?

**Comment by brammool** [comment]
> I am enjoying to use diff mode.
> But I am wondering why the wrap option always forced to be off.
> 
> I know its reason is to properly align text between diff windows,
> but it is more important for some users (including me) to always wrap lines and see all text.
> We have to reset the option every time in diff windows (or set autocmd).
> 
> Is it possible and reasonable to remain the wrap option in diff mode?

It would be possible, but it's really a new display mode, thus it will
be work to implement.  Obviously a side-by-side view can only be in sync
at one point in the window (top or middle).  Or we would have to insert
filler lines where needed, there already is the "filler" item in
'diffopt'.

-- 
hundred-and-one symptoms of being an internet addict:
133. You communicate with people on other continents more than you
     do with your own neighbors.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by rickhowe** [comment]
> It would be possible, but it's really a new display mode, thus it will be work to implement. Obviously a side-by-side view can only be in sync at one point in the window (top or middle). Or we would have to insert filler lines where needed, there already is the "filler" item in 'diffopt'.

Yes, to properly align text even when the wrap is on, more 'filler' lines might be necessary.
But it is OK if the alignment goes wrong in this case.
The issue here is that there is no choice and need to set wrap option again every time.

How do you think , like 'filler', to add another option in diffopt? Below is an example  to set nowrap or not:

```
--- optiondefs0.h	2020-08-07 18:48:25.490695800 +0900
+++ optiondefs.h	2020-08-07 18:48:42.611993900 +0900
@@ -847,7 +847,7 @@
 								     |P_NODUP,
 #ifdef FEAT_DIFF
 			    (char_u *)&p_dip, PV_NONE,
-			    {(char_u *)"internal,filler,closeoff",
+			    {(char_u *)"internal,filler,closeoff,nowrap",
 								(char_u *)NULL}
 #else
 			    (char_u *)NULL, PV_NONE,
```
```
--- diff0.c	2020-06-10 10:13:22.407273800 +0900
+++ diff.c	2020-08-07 18:53:43.721994500 +0900
@@ -36,8 +36,9 @@
 #define DIFF_HIDDEN_OFF	0x100	// diffoff when hidden
 #define DIFF_INTERNAL	0x200	// use internal xdiff algorithm
 #define DIFF_CLOSE_OFF	0x400	// diffoff when closing window
+#define DIFF_NOWRAP	0x800	// nowrap
 #define ALL_WHITE_DIFF (DIFF_IWHITE | DIFF_IWHITEALL | DIFF_IWHITEEOL)
-static int	diff_flags = DIFF_INTERNAL | DIFF_FILLER | DIFF_CLOSE_OFF;
+static int	diff_flags = DIFF_INTERNAL | DIFF_FILLER | DIFF_CLOSE_OFF | DIFF_NOWRAP;
 
 static long diff_algorithm = 0;
 
@@ -1454,9 +1455,12 @@
     if (!wp->w_p_diff)
 	wp->w_p_crb_save = wp->w_p_crb;
     wp->w_p_crb = TRUE;
-    if (!wp->w_p_diff)
-	wp->w_p_wrap_save = wp->w_p_wrap;
-    wp->w_p_wrap = FALSE;
+    if (diff_flags & DIFF_NOWRAP)
+    {
+	if (!wp->w_p_diff)
+	    wp->w_p_wrap_save = wp->w_p_wrap;
+	wp->w_p_wrap = FALSE;
+    }
 # ifdef FEAT_FOLDING
     if (!wp->w_p_diff)
     {
@@ -1517,8 +1521,11 @@
 		    wp->w_p_scb = wp->w_p_scb_save;
 		if (wp->w_p_crb)
 		    wp->w_p_crb = wp->w_p_crb_save;
-		if (!wp->w_p_wrap)
-		    wp->w_p_wrap = wp->w_p_wrap_save;
+		if (diff_flags & DIFF_NOWRAP)
+		{
+		    if (!wp->w_p_wrap)
+			wp->w_p_wrap = wp->w_p_wrap_save;
+		}
 #ifdef FEAT_FOLDING
 		free_string_option(wp->w_p_fdm);
 		wp->w_p_fdm = vim_strsave(
@@ -2245,6 +2252,11 @@
 	    p += 8;
 	    diff_flags_new |= DIFF_CLOSE_OFF;
 	}
+	else if (STRNCMP(p, "nowrap", 6) == 0)
+	{
+	    p += 6;
+	    diff_flags_new |= DIFF_NOWRAP;
+	}
 	else if (STRNCMP(p, "indent-heuristic", 16) == 0)
 	{
 	    p += 16;
```
